tests/extmod_hardware/machine_pwm.py

```diff
@@ -0,0 +1,108 @@
+try:
+    from machine import time_pulse_us, Pin, PWM
+except ImportError:
+    print("SKIP")
+    raise SystemExit
+
+freq_margin_per_thousand = 0
+duty_margin_per_thousand = 0
+timing_margin_us = 5
+
+# Configure pins based on the target.
+if "esp32" in sys.platform:
+    pwm_pin = 4
+    pulse_pin = 5
+    freq_margin_per_thousand = 2
+    duty_margin_per_thousand = 1
+    timing_margin_us = 20
+elif "esp8266" in sys.platform:
+    pwm_pin = 4
+    pulse_pin = 5
+    duty_margin_per_thousand = 3
+    timing_margin_us = 50
+elif "mimxrt" in sys.platform:
+    pwm_pin = "D0"
+    pulse_pin = "D1"
+elif "rp2" in sys.platform:
+    pwm_pin = "GPIO0"
+    pulse_pin = "GPIO1"
+elif "samd" in sys.platform:
+    pwm_pin = "D0"
+    pulse_pin = "D1"
+else:
+    print("Please add support for this test on this platform.")
+    raise SystemExit
+
+
+def test_freq_duty(pulse_in, pwm, freq, duty_u16):
+    print("freq={:<5} duty_u16={:<5} :".format(freq, duty_u16), end="")
+
+    # Check configured freq/duty_u16 is within error bound.
+    freq_error = abs(pwm.freq() - freq) * 1000 // freq
+    duty_error = abs(pwm.duty_u16() - duty_u16) * 1000 // (duty_u16 or 1)
+    print("", freq_error <= freq_margin_per_thousand or (freq, pwm.freq()), end="")
+    print("", duty_error <= duty_margin_per_thousand or (duty_u16, pwm.duty_u16()), end="")
+    print(" :", end="")
+
+    # Calculate expected timing.
+    expected_total_us = 1_000_000 // freq
+    expected_high_us = expected_total_us * duty_u16 // 65535
+    expected_low_us = expected_total_us - expected_high_us
+    expected_us = (expected_low_us, expected_high_us)
+    timeout = 2 * expected_total_us
+
+    # Wait for output to settle.
+    time_pulse_us(pulse_in, 0, timeout)
+    time_pulse_us(pulse_in, 1, timeout)
+
+    if duty_u16 == 0 or duty_u16 == 65535:
+        # Expect a constant output level.
+        no_pulse = (
+            time_pulse_us(pulse_in, 0, timeout) < 0 and time_pulse_us(pulse_in, 1, timeout) < 0
+        )
+        if expected_high_us == 0:
+            # Expect a constant low level.
+            print("", 0, no_pulse and pulse_in() == 0)
+        else:
+            # Expect a constant high level.
+            print("       ", 1, no_pulse and pulse_in() == 1)
+    else:
+        # Test timing of low and high pulse.
+        n_averaging = 10
+        for level in (0, 1):
+            t = 0
+            time_pulse_us(pulse_in, level, timeout)
+            for _ in range(n_averaging):
+                t += time_pulse_us(pulse_in, level, timeout)
+            t //= n_averaging
+            expected = expected_us[level]
+            print("", level, abs(t - expected) <= timing_margin_us or (expected, t), end="")
+        print()
+
+
+def test():
+    pulse_in = Pin(pulse_pin, Pin.IN)
+    pwm = PWM(pwm_pin)
+
+    for freq in (50, 100, 500, 1_000, 2_000, 5_000, 10_000):
+        pwm.freq(freq)
+        for duty in (0, 10, 25, 50, 75, 90, 100):
+            duty_u16 = duty * 65535 // 100
```

Comment:
Must be `duty_u16 = duty * 65536 // 100` to get a high pulse at all devices.

Reply:
But that fails with the actual ESP32 firmware. With PR #10854 it works properly. That PR is  major improvement for PWM. Even with the current implementation of PWM for ESP32, the output is NOT constant high at duty_u16=65535, but time_pulse_us() may be too slow at ESP32 to sample it.

Reply:
That's not true according to the docs, and the rp2 implementation.  Both of those have 65535 as the upper limit.

This definitely needs to be clarified!

Reply:
> This definitely needs to be clarified!

The docs have to be updated. Besides ESP32 all ports accept 65536 as setting the output constant high and 0 for setting the output constant low. Depending on the dividers and frequencies, there may be no output pulse at e.g. 65535 (or 1).

Reply:
A duty of 0 should definitely mean a duty of 0%, ie constant low output.

Then if the duty range is 16-bits, the maximum you can specify is 65535.  And so that must be a constant high output.  (For example, what if you are storing duty values in 2 bytes of data, you need to be able to specify the whole range in 2 bytes.)

Reply:
But doing it that way there is no 50% or 25% duty. Just numbers close to it. We can of course define 65535 as the value of "constant on", but keep the ratio as duty_u16/65536.

Reply:
There's also no perfectly accurate 10%, whether the division is /65535 or /65536.

With /65535 there is an accurate 20%, 40%, 60% and 80%.

But if the resolution of the underlying hardware PWM counter is <= 65535 (eg on rp2), then it should always be possible to select all of the available hardware duty cycles, including an accurate 50% (with /65535).  Eg on rp2 a duty_u16 of 32768 should give you exactly 50%, due to integer rounding when calculating the hardware counter limit.